.github/workflows/build.yml

```diff
@@ -0,0 +1,227 @@
+    outputs:
+      filepath: ${{ steps.generate-metadata-file.outputs.filepath }}
+    steps:
+      - uses: actions/checkout@v3
+      - name: Generate package metadata
+        id: generate-metadata-file
+        uses: hashicorp/actions-generate-metadata@v1
+        with:
+          version: ${{ needs.get-product-version.outputs.product-version }}
+          product: ${{ env.PKG_NAME }}
+      - uses: actions/upload-artifact@v2
+        with:
+          name: metadata.json
+          path: ${{ steps.generate-metadata-file.outputs.filepath }}
+
+  build-ui:
+    runs-on: ubuntu-latest
+    name: Build UI
+    needs: get-product-version
+    steps:
+      - uses: actions/checkout@v3
+      - name: Set up node and yarn
+        uses: actions/setup-node@v3
+        with:
+          node-version: 14
+          cache: yarn
+          cache-dependency-path: ui/yarn.lock
+      - name: Build UI
+        run: |
+          export PKG_VERSION=${{ needs.get-product-version.outputs.product-version }}
+          export PKG_BINARY_TYPE=${{ needs.get-product-version.outputs.pkg-metadata }}
+          export PKG_COPYRIGHT_YEAR=$(git show -s --format=%cd --date=format:%Y HEAD)
+          export PKG_NAME=${{ needs.get-product-version.outputs.pkg-name }}
+          echo "${PKG_NAME} version is ${PKG_VERSION}"
+          echo "${PKG_NAME} binary type is ${PKG_BINARY_TYPE}"
+          echo "${PKG_NAME} copyright year is ${PKG_COPYRIGHT_YEAR}"
+          cd ui && make && cd ..
+      - name: Save assets to cache
+        uses: actions/cache/save@v3
+        with:
+          path: ui/dist/assets
+          key: ui_assets_${{ github.sha }}
+
+  build-assets:
+    # FIXME: can releases be tagged as per CRT standards (ie at the end of the pipeline)?
+    runs-on: ubuntu-latest
+    name: Build Assets
+    needs:
+      - get-go-version
+      - build-ui
+    steps:
+      - uses: actions/checkout@v3
+      - uses: actions/setup-go@v3
+        with:
+          go-version: ${{ needs.get-go-version.outputs.go-version }}
+      - name: Restore ui assets
+        uses: actions/cache/restore@v3
+        with:
+          path: ui/dist/assets
+          key: ui_assets_${{ github.sha }}
+          fail-on-cache-miss: true
+          enableCrossOsArchive: true
+      - name: Build assets
+        run: |
+          make tools
+          go install github.com/kevinburke/go-bindata/go-bindata@v3.24.0+incompatible
+          make bin/crt-assets
+      - name: Save assets to cache
+        uses: actions/cache/save@v3
+        with:
+          path: internal/assets
+          key: build_assets_${{ github.sha }}
+
+  build-waypoint:
+    needs:
+      - get-product-version
+      - get-go-version
+      - build-assets
+      - build-ui
+    strategy:
+      matrix:
+        pkg-name: [ "waypoint", "waypoint-entrypoint" ]
+        goos: [ linux ]
+        goarch: [ "386", "amd64", "arm", "arm64" ]
+        # FIXME: can we confirm which os/arch combos we require for this pipeline?
```

Comment:
@waypoint - Please confirm which os/arch combinations are necessary for this pipeline. As of now we have:

1. Linux/386 (+entrypoint)
2. Linux/amd64 (+entrypoint)
3. Linux/arm (+entrypoint)
4. Linux/arm64 (+entrypoint)
5. Darwin/amd64 (+entrypoint)
6. Darwin/arm64 (+entrypoint)
7. Windows/386
8. Windows/amd64

Reply:
Correct: https://hashicorp.slack.com/archives/C013QT1KG9W/p1679517514985589